### PR TITLE
Implement dynamic community selection for global search

### PR DESCRIPTION
**Pull request by bryanlimy**
## Description

Replacing the previous PR https://github.com/microsoft/graphrag/pull/1219

Implement dynamic community selection to improve global search result. 

Currently, we select the community reports by specifying a specific community level, prior to the map-reduce operation, while fast, this operation could eliminate useful community reports (i.e. at lower levels) and/or include reports that are not informative to the query.

Here, we take advantage of the hierachical structure of the indexed dataset. We first ask the LLM to rate how relevant each level 0 community is with respect to the user query, we then traverse down the child node(s) if the current community report is deemed relevant. Ultimately reducing the number of community reports that get passed to the map-reduce operation in global search.

- updated `SearchResult` to store the break down of LLM calls, prompt tokens and output tokens, so that we can compute the token costs separately by model.
- Updated DRIFT search to record LLM calls, prompt and output tokens.

## Related Issues

[Reference any related issues or tasks that this pull request addresses.]

## Proposed Changes

[List the specific changes made in this pull request.]

## Checklist

- [x] I have tested these changes locally.
- [x] I have reviewed the code changes.
- [x] I have updated the documentation (if necessary).
- [ ] I have added appropriate unit tests (if applicable).

## Additional Notes

[Add any additional notes or context that may be helpful for the reviewer(s).]

### CODE REVIEW

File: graphrag/query/structured_search/drift_search/action.py [hunk 1]
```diff
@@ -39,7 +39,11 @@ def __init__(
         self.follow_ups: list[DriftAction] = (
             follow_ups if follow_ups is not None else []
         )
-        self.metadata: dict[str, Any] = {}
+        self.metadata: dict[str, Any] = {
```
Comment (bryanlimy) [inline graphrag/query/structured_search/drift_search/action.py:42]: I would like to maintain the existing `self.metadata` to record token usage. But if the same `DriftAction` object is being used for different queries, then we will need to reset these counters per query.

File: graphrag/query/context_builder/dynamic_community_selection.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        self.llm = llm
+        self.token_encoder = token_encoder
+        self.rate_query = rate_query
+        self.num_repeats = num_repeats
+        self.use_summary = use_summary
+        self.threshold = threshold
+        self.keep_parent = keep_parent
+        self.max_level = max_level
+        self.semaphore = asyncio.Semaphore(concurrent_coroutines)
+        self.llm_kwargs = {"temperature": 0.0, "max_tokens": 2000}
```
Comment (ha2trinh) [inline graphrag/query/context_builder/dynamic_community_selection.py:50]: should the llm_kwargs be configurable (with this as default values?)

File: graphrag/query/context_builder/rate_relevancy.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import numpy as np
+import tiktoken
+
+from graphrag.llm.openai.utils import try_parse_json_object
+from graphrag.query.llm.base import BaseLLM
+from graphrag.query.llm.text_utils import num_tokens
+
+log = logging.getLogger(__name__)
+
+RATE_QUERY = """
```
[the remark this answers is not in the document]
Comment (bryanlimy) [reply graphrag/query/context_builder/rate_relevancy.py]: updated

**Review by natoverse — DISMISSED** [review]

**Comment by bryanlimy** [comment]
Merged in https://github.com/microsoft/graphrag/pull/1396

**Pull request closed without merging**